### PR TITLE
[BREAKING] Compile {{fa-icon}} statically where possible.

### PR DESCRIPTION
**Pull request by cibernox**
I've been playing with AST transforms and I wanted to do it on a real world addon on a project that might actually benefit for real of the idea.

I've created an AST transform that check if all arguments passed to `{{fa-icon}}` are static, and if they are, it just transforms those occurences to `<i aria-hidden="true" class="fa fa-whatever"></i>`). If any attribute is dynamic, it leaves it untouched.

I also use the transforms to decided wether or not all invocations in an app are static, and what values have been used.

This could be used for, if all invocations are static, bundle only the icons that the app actually uses.

### CODE REVIEW

**Comment by cibernox** [comment]
Supported usages so far:

  ```hbs
  1. {{fa-icon icon="credit-card"}}
  2. {{fa-icon icon="credit-card" fixedWidth=true}}
  3. {{fa-icon "credit-card"}}
  4. {{fa-icon "credit-card" ariaLabel="Select card"}}
  5. {{fa-icon "credit-card" ariaHidden=false}}
  6. {{fa-icon "credit-card" class="custom-class"}}
  7. {{fa-icon "credit-card" tagName="span"}}
  8. {{fa-icon "credit-card" size="lg"}}
  9. {{fa-icon "credit-card" size=2}}
  ```
  becomes
  ```hbs
  1. <i aria-hidden="true" class="fa fa-credit-card"></i>
  2. <i aria-hidden="true" class="fa fa-credit-card fa-fw"></i>
  3. <i aria-hidden="true" class="fa fa-credit-card"></i>
  4. <i aria-hidden="true" aria-label="Select card" class="fa fa-credit-card"></i>
  5. <i class="fa fa-credit-card"></i>
  6. <i aria-hidden="true" class="fa fa-credit-card custom-class"></i>
  7. <span aria-hidden="true" class="fa fa-credit-card"></span>
  8. <i aria-hidden="true" class="fa fa-credit-card fa-lg"></i>
  9. <i aria-hidden="true" class="fa fa-credit-card fa-2x"></i>
  ```

Even bound usages like `{{fa-icon icon=boundValue}}` can be transformed into `<i class="fa fa-{{boundValue}}"></i>`. So far the ONLY usage that I've found that cannot be transformed in buildTime is `{{fa-icon "some-icon" tagName=boundValue}}`, would be pretty rare (note that `{{fa-icon "some-icon" tagName="span"}}` is static and can be transformed)

**Comment by buschtoens** [comment]
This is already looking so sexy. Amazing work. 👏

**Comment by cibernox** [comment]
Ok, I've managed to convert nearly 100% of the functionality of the addon into a build-time AST transform.

Essentially users should be able to update without noticing anything with a couple exceptions. There is two particular features that are a PITA to implement with AST and since they were just small conveniences, I'd rather remove those features than to make the transform too complicated.

Those features are:
- Dynamic tagName: `{{fa-icon "credit-card" tagName=someBoundValue}}`. This is a bad idea either way.
- Allow the icon name repeat the `fa-` prefix on the icon-name: `{{fa-icon "fa-credit-card"}}`. This seemed redundant, the intended usage is `{{fa-icon "credit-card"}}`.
- Allow stack and size to be strings ending in `x`: `{{fa-icon "fa-credit-card" size="1x"}}`.  You can pass `"1"`, `1` as before, but not `"1x"`. 

That is pretty much it. Now I'm going to work in gathering the used icons and bundle just those that are used.
Also, I want to run some benchmarks on this.

**Comment by cibernox** [comment]
For now, the results are promissing:

```
#### BEFORE
File sizes:
 - dist/assets/dummy-79a9f017c51f5f4b0570e328952f632a.js: 109.85 KB (6.23 KB gzipped)
 - dist/assets/vendor-1c8ce4f720fc5b44b422432b8939e629.js: 608.08 KB (165.35 KB gzipped)
 - dist/assets/vendor-2cc3ab6c0c8b8b7103a3a8673ec9eb56.css: 30.18 KB (6.89 KB gzipped)

#### AFTER
File sizes:
 - dist/assets/dummy-396a64bbe42d7d89ad8f2561dec04fa1.js: 3.97 KB (1.26 KB gzipped)
 - dist/assets/vendor-3c28bba004b9b91de76abcd06b517395.js: 604.58 KB (164.55 KB gzipped)
 - dist/assets/vendor-2cc3ab6c0c8b8b7103a3a8673ec9eb56.css: 30.18 KB (6.89 KB gzipped)
```

**Comment by buschtoens** [comment]
I can't thank you enough for this awesome PR.  
If I can make it to EmberFest, I'll definitely owe you a beer! :beers: :smile:

**Comment by cibernox** [comment]
I did some preliminary benchmark and the speed improvement is on the line with what I expected

Before: average time of ~175ms
screen shot 2017-08-24 at 10 01 50

After: Average time of ~40ms 
screen shot 2017-08-24 at 10 00 43

Between 400% and 500% faster.

**Comment by cibernox** [comment]
Update:

I managed to pass the stylesheets though postcss to remove the classes of unused icons. It's still slow and it doesn't work in rebuilds, but it does strip css classes of unused icons. Using just 6 icons the results are:

Before:
`dist/assets/vendor-2cc3ab6c0c8b8b7103a3a8673ec9eb56.css: 30.18 KB (6.89 KB gzipped)`

After: 
`dist/assets/vendor-617a7955da24bfd27d5d0339565889a0.css: 3.48 KB (1.13 KB gzipped)
`

**Comment by cibernox** [comment]
This PR is already functional. It would be asomeome if someone can install this on their app to betatest any edge case (I'm sure there will be some) to get it polished before merging

**Comment by jrjohnson** [comment]
Attempted to use this in an addon which provides components and it fails with the error `Compile Error: fa-icon is not a helper` when `{{fa-icon}}` is used from within a component inside the addon.

Some steps to reproduce:
```
ember addon component-provider --yarn
ember generate test-icon
ember install ember-font-awesome@cibernox/ember-font-awesome#transform-fa-icon-statically
```
Then add to `test-icon.hbs`
```hbs
{{fa-icon 'plane'}}
```
And call it in `application.hbs`
```hbs
{{test-icon}}
```
`ember serve` and visit and you will see nothing and the error is in the console.

**Comment by jrjohnson** [comment]
Doesn't support `{{fa-icon 'click' click=(action 'clickClock')}}`
Gives the error `Found unknown hash option named "click"`

**Comment by jrjohnson** [comment]
Fails when the title is passed another helper:
`{{fa-icon 'plus' title=(concat 'a' 'b')}}`
in our app this is almost always going to be the `t` helper provided by ember-i18n, but `concat` is a cleaner test case.

**Comment by cibernox** [comment]
Nice, those are the kind of test cases I was looking for

**Comment by cibernox** [comment]
@jrjohnson I've fixed most, if not all of the corner cases. I haven't tried the scenario where the addon is used not from within application templates but addon templates. I'll try later or tomorrow.

File: lib/ast-transform.js [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+          }
+          break;
+        case 'stack':
+          stack = pair.value;
+          break;
+        case 'click':
+          attrs.push(buildAttr(b, 'onclick', pair.value));
+          break;
+        default:
+          console.warn(`Found unknown hash option named "${pair.key}"`);
```
Comment (jrjohnson) [inline lib/ast-transform.js]: This warning should include more details. At the very least it should include `ember-font-awesome` if it isn't possible to grab the file and line. Otherwise a warning like 
`Found unknown has option names "classNames"` is hard to track down in a codebase that might have many classNames declarations.

Comment (cibernox) [reply lib/ast-transform.js]: Added to the queue of changes. Initially I did this for myself, but I think it would be useful for getting good bug reports.

**Comment by cibernox** [comment]
@jrjohnson I've made some changes and now it can be used inside other addons. 

However, since it performs transforms on templates, it has to be moved from `devDependencies` to `dependencies`, just like `ember-cli-htmlbars` does.

**Comment by jrjohnson** [comment]
I was just wondering if that was the case. I just tested and it works great inside an addon. Is there a way to present the same kind of dire warning that `ember-cli-htmlbars` presents to let users know they should move this dependency?

File: tests/dummy/app/templates/index.hbs [hunk 2]
```diff
@@ -13,6 +13,7 @@
       <li>{{fa-icon icon='angellist' fixedWidth=true}} <span class="name">angellist</span></li>
       <li>{{fa-icon icon='angle-double-down' fixedWidth=true}} <span class="name">angle-double-down</span></li>
       <li>{{fa-icon icon='angle-double-left' fixedWidth=true}} <span class="name">angle-double-left</span></li>
+      {{!--
```
Comment (jrjohnson) [inline tests/dummy/app/templates/index.hbs]: I'm guessing you commented these out while testing and now this should be removed.

File: lib/ast-transform.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            classContent = appendToContent(b, `fa-${pair.value.value}x`, classContent);
+          } else if (isNaN(parseInt(pair.value.value), 10)) {
+            classContent = appendToContent(b, `fa-${pair.value.value}`, classContent);
+          } else {
+            classContent = appendToContent(b, `fa-${pair.value.value}x`, classContent);
+          }
+          break;
+        case 'pull':
+          if (isDynamic(pair.value)) {
+            classContent = appendToContent(b, 'fa-pull-', classContent);  // This is going to be wrong
```
Comment (jrjohnson) [inline lib/ast-transform.js]: what does `// This is going to be wrong` mean?

Comment (cibernox) [reply lib/ast-transform.js]: Leftovers from when that didn't work. Now it does.

**Comment by cibernox** [comment]
@jrjohnson I don't see an easy a way, because in reality this is only needed when the addon is a dependency of another addon, not when it's a direct dependency of the app.

I'll ask the wise elders of there is a way, but I wouldn't hold this on an uncommon situation (fa-icon as a 2nd degree dependency)

**Comment by jrjohnson** [comment]
Alright - if you need some beer to help ply the wise elders or they give you a quest that required a party let me know!

I looked around in `ember-cli-htmlbars` and I couldn't find anything that throws that nice error message so it must be somewhere else. We can just make sure to document this condition before beta becomes release.

**Comment by cibernox** [comment]
Ok, I'm going to merge and release as 4.0.0-alpha.0, and add a visible disclaimer in the readme.
